rfcs/2021-10-15-9572-accept-datadog-traces.md

```diff
@@ -0,0 +1,216 @@
+[trace-agent]: https://github.com/DataDog/datadog-agent/blob/0a19a75/pkg/trace/event/processor.go#L47-L91
+[short description of APM events]: https://github.com/DataDog/datadog-agent/blob/e081bed/pkg/trace/event/doc.go
+
+## Cross cutting concerns
+
+N/A
+
+## Scope
+
+### In scope
+
+* Ingest traces from the trace-agent in the `datadog_agent` source
+* Send traces to the Datadog trace endpoint through a new `datadog_trace` sink
+* Basic operation on traces: filtering, routing
+* Pave the way for OpenTelemetry traces
+
+### Out of scope
+
+* Profiling data, but as the trace-agent only proxies profiling data, the same behaviour can be implemented rather
+  quickly in Vector.
+* Debugger logs can be already diverted to Vector (untested but it should work as Vector supports datadog logs and there
+  is an config option to explicitly configure the debugger log destination)
+* Metrics emitted by the trace-agent (they could theoretically be received by Vector by a statsd source, but the host
+  used by the trace-agent is derived from the local config to programmatically discover the main agent, thus there is no
+  existing knob to force the trace agent to send metrics to a custom dogstatsd host)
+* Span extraction, filtering
+* Other sources & sinks for traces than `datadog_agent` source & `datadog_trace` sink
+
+## Pain
+
+Vector does not support any traces (full json representation may be ingested as log event) at the moment and it is a
+key part of observability. Therefore, users cannot use Vector for the business-level user cases on trace data, like
+cost control and reduction, redacting PII, routing, and more.
+
+## Proposal
+
+### User Experience
+
+* User will be able to ingest traces from the trace agent
+  * Vector config would then consist of: `datadog_agent` source -> some filtering/enrichment transform ->
+    `datadog_trace` sink
+  * Datadog trace agent can be configured to send traces to any arbitrary endpoint using `apm_config.apm_dd_url` [config
+    key]
+* This change is a pure addition to Vector, there will be no impact on existing feature
+
+[config key]: https://github.com/DataDog/datadog-agent/blob/34a5589/pkg/config/apm.go#L61-L87
+
+### Implementation
+
+To keep vector-core as generic as possible, the first implementation will decode datadog traces as `LogEvent`, the
+resulting event will be deeper than usual but this should not be a problem. To keep track that those `LogEvent`s are
+traces, we could add a simple `Option<bool>` to the `EventMetadata` struct. However this may not be immediately
+required. While being naive this does not predates future vector evolution towards event traits by not adding a new
+concrete type into the `Event` enum.
+
+Upcoming [work][schema-rfc] on having the ability to validate a `LogEvent` against a schema would provide a
+nice way (with the performance question) of ensuring that a `datadog-traces` sinks would receive a properly structured
+`LogEvent`.
```

Comment:
My biggest concern with this approach is correct interoperability between tracing protocols. This same concern extends to processing methods like tail-based sampling, which needs to understand how spans are correlated across different tracing protocols.

Another way to explain this is through the lens of metrics. The interoperability concern is similar. Imagine representing metrics as logs and then implementing correct interoperability between metrics protocols. It would not be possible.

By introducing a `trace` event type that normalizes around a specific structure we force all of your tracing components to properly map data and work for all tracing protocols.

Does that make sense? And would the approach outlined here preclude that?

cc @lukesteensen @blt @jszwedko

Reply:
@prognant we discussed this offline. See [this thread](https://dd.slack.com/archives/C01UQ5ZPV6K/p1639159935222300). I am in favor of this now, but would like the RFC to reflect that we intend to use the upcoming schema feature for interoperability.